Quickstart javascript express web (#2449)
FusionAuth/fusionauth-site: astro/src/content/quickstarts/quickstart-javascript-express-web.mdx

cta: EmailListCTA
---
import Aside from '/src/components/Aside.astro';
import LoginAfter from '../../diagrams/quickstarts/login-after.astro';
import LoginBefore from '../../diagrams/quickstarts/login-before.astro';
import RemoteCode from '/src/components/RemoteCode.astro';

In this quickstart you are going to build an application with Express and integrate it with FusionAuth.
You'll be building it for [ChangeBank](https://www.youtube.com/watch?v=CXDxNCzUspM), a global leader in converting dollars into coins.
It'll have areas reserved for users who have logged in as well as public facing sections.

The docker compose file and source code for a complete application are available at
[https://github.com/FusionAuth/fusionauth-quickstart-javascript-express-web](https://github.com/FusionAuth/fusionauth-quickstart-javascript-express-web).

## Prerequisites
* [Node LTS](https://nodejs.org/en/download)
* [Docker](https://www.docker.com): The quickest way to stand up FusionAuth. (There are [other ways](/docs/v1/tech/installation-guide/)).
* [git](https://git-scm.com/downloads/): Not required but recommended if you want to track your changes.

> * [express 4.16+](https://expressjs.com/): Web Framework for Node.js 

Author: great point!

```
docker compose up -d
```

This will start three containers, once each for FusionAuth, Postgres, and Elastic.

Here you are using a bootstrapping feature of FusionAuth, called [Kickstart](/docs/v1/tech/installation-guide/kickstart).
When FusionAuth comes up for the first time, it will look at the `kickstart/kickstart.json` file and
configure FusionAuth to a certain initial state.

<Aside type="note">
If you ever want to reset the FusionAuth system, delete the volumes created by Docker Compose by
executing `docker compose down -v`, then re-run `docker compose up -d`.
</Aside>

FusionAuth will be configured with these settings:
* Your client Id is `e9fdb985-9173-4e01-9d73-ac2d60d1dc8e`.
* Your client secret is `2HYT86lWSAntc-mvtHLX5XXEpk9ThcqZb4YEh65CLjA-not-for-prod`.
* Your example username is `richard@example.com` and the password is `password`.
* Your admin username is `admin@example.com` and the password is `password`.
* The base URL of FusionAuth `http://localhost:9011/`.

You can log into the [FusionAuth admin UI](http://localhost:9011/admin) and look around if you want, but with Docker/Kickstart you don’t need to.

### Setup Application

> In this section, you'll set up a basic Express application with a single page.

Reviewer: This app has a number of pages, correct?

Create a new directory to hold your application, and go into it.

```shell
mkdir changebank
cd changebank
```

Create a file called `package.json` with the below dependencies and scripts. Feel free to update the `author` or other attributes as needed. 

The `@fusionauth/typescript-client` will be used to exchange the OAuth Code for Access Token using PKCE, and also to retrieve the User using JWT.

> `package.json`

Reviewer: I'd rather not have the filename here like this. That is inconsistent with other quickstarts.

Reviewer: Or we can change it across all of them, but that's probably a clean up task.

<Aside type="note">
In this file you will use the [pkce-challenge](https://www.npmjs.com/package/pkce-challenge) package to generate a Proof Key for Code Exchange (PKCE) challenge pair. Read more about [PKCE](/articles/oauth/modern-guide-to-oauth).
</Aside>

Include the below code into `index.ts`, so that the homepage can be displayed when a user is not logged in. If a user is logged in this route will redirect to the `/account` route.

<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="homepage"/>

### Login Route

In the ChangeBank application there is a `Login` button which will open the `/login` route when clicked. This route redirect them to a login page managed by FusionAuth.

<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="login"/>

> - `client_id`: Your `applicationId`

Reviewer: You refer to it as a client Id above. This is the only time you refer to this as `applicationId`. I think we should refer to it as client Id.

- `response_type`: `code` since you are using the Authorization Code grant
- `redirect_uri`: Where your FusionAuth will redirect back to during the OAuth grant. In this example, this is our route `oauth-redirect`, which you'll build next.
- `state`: A random string created in your homepage route and stored in the cookie as `stateValue`
- `code_challenge`: Created in the homepage route using `pkce-challenge` and stored in a cookie as `challenge`, this is one half of the PKCE 

### OAuth Redirect

When FusionAuth redirects back to your Express application, it will return to your `/oauth-redirect` route. In this response it will have two query parameters that you need to further examine: `code` and `state`. 

You'll need to do the following:

1. Check that the `stateValue` stored in the cookie matches the `state` query parameter.

> 1. Make a request to the FusionAuth token endpoint to exchange the authorization `code` received after the user authenticates at FusionAuth and a `verifier` from our `userSession` cookie for an access token.

Reviewer: I don't understand this sentence.

1. Retrieve the user by passing the access token to FusionAuth.
1. Store access token details in the `userToken` cookie.
1. Store retrieved user in `userDetails` cookie.
1. Redirect to `/account` route.

<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="oauth-redirect"/>

### Account Route

The `/account` page represents what a user sees when they’re logged into their ChangeBank account. A user that isn’t logged in who tries to access this page should just be taken to the home page.

<Aside type="note">

> In a production application, you would validate the token signature, make sure it wasn't expired, and attempt to refresh it if it were expired.

Reviewer: Does this apply now that you are doing the signature check?

</Aside>

<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="account"/>

### Make Change

> Two functions are needed to accept both GET and POST requests on the `/make-change` route. The `GET` verifies that the user is logged in, then servers the HTML file required for the page. The `POST` takes the amount form value as a body parameter and uses the it to calculate how many quarters, dimes, nickels and pennies to return. If `total` is not a valid dollar value or cannot be converted an `error` message is returned. The return string `message` is received by asyncronously and updates the `change-message` or `error-message` div accordingly.

Reviewer: ```suggestion
Two functions are needed to accept both `GET` and `POST` requests on the `/make-change` route. The `GET` verifies that the user is logged in, then serves the HTML file required for the page. The `POST` takes the amount form value as a body parameter and uses the it to calculate how many quarters, dimes, nickels and pennies to return. If `total` is not a valid dollar value or cannot be converted an `error` message is returned. The return string `message` is received by asynchronously and updates the `change-message` or `error-message` div accordingly.
```

Add the code below to `index.ts`.

<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="make-change"/>

### Logout Route

<Aside type="tip">
Click the `Logout` button and watch the browser first go to FusionAuth to log out the user, then return to your home page.
</Aside>

The last step is to implement logout. When you log a user out of an application, you'll take them to 

> FusionAuth's `/oauth2/logout` endpoint. FusionAuth calls the logout endpoint (killing the express session), then kills the FusionAuth session, then redirects to the logout endpoint. This endpoint deletes any cookies that your application created, clearing the user's session.

Reviewer: I don't understand this. Why do we call the logout endpoint two times? Is it because the first time kills the session and the second time kills cookies? I don't get it.

Author: @mooreds right now it is sharing where FusionAuth redirects back with the application calling the same. Would it be cleaner to add an additional endpoint called `/oauth2/logout` so there is less confusion? 

Or I could even change the button to call `${fusionAuthURL}/oauth2/logout?client_id=${clientId}` directly instead, I just didn't want to put any logic into the html.

Author: I went for the former and updated.

<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="logout"/>

In the below code, you check if the current user session has any information about the user and redirect to the FusionAuth server to initiate the logout on the server. When this returns, the user session cookie will now be empty and you can therefore forward back to your homepage. If you didn't add this check, you would be stuck in an infinite loop between your Express server and FusionAuth.

```ts
if (userSession) {
    res.redirect(302, `${fusionAuthURL}/oauth2/logout?client_id=${clientId}`);
} else {
    res.redirect(302, '/')
}
```

> 

Reviewer: What do I do now? I think I run `npm run dev` but please tell me to do so.

</Aside>

FusionAuth will be configured with these settings:
* Your client Id is `e9fdb985-9173-4e01-9d73-ac2d60d1dc8e`.
* Your client secret is `2HYT86lWSAntc-mvtHLX5XXEpk9ThcqZb4YEh65CLjA-not-for-prod`.
* Your example username is `richard@example.com` and the password is `password`.
* Your admin username is `admin@example.com` and the password is `password`.
* The base URL of FusionAuth `http://localhost:9011/`.

You can log into the [FusionAuth admin UI](http://localhost:9011/admin) and look around if you want, but with Docker/Kickstart you don’t need to.

### Setup Application

In this section, you'll set up a basic Express application with a single page.

Create a new directory to hold your application, and go into it.

```shell
mkdir changebank
cd changebank
```

To keep track of changes, initialize your directory with git.

```shell
git init
```

Create a file called `.gitignore` so that you can exclude files from your git repo.

`.gitignore`
<RemoteCode url={frontmatter.coderoot + "/complete-application/.gitignore"}
  lang="text" />

Create a file called `package.json` with the below dependencies and scripts.

> `package.json`

Reviewer: How is fusionauth-client used?

1. Make sure to check that the `stateValue` that we stored in our cookie matches the `state` query parameters
1. Request to the FusionAuth Token endpoint to exchange the authorization `code` returned from the Authorize endpoint and a `verifier` from our `userSession` cookie for an access token.
1. Retrieve the user by passing the access token to FusionAuth. (This method does not use an API key, instead it uses a JSON Web Token (JWT) for authentication).
1. Store access token details in `userToken` cookie.
1. Store retrieved user in `userDetails` cookie (note for our example this is not a secure cookie and JS can access).
1. Redirect to `/account` route.

`index.ts`
<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="oauth-redirect"/>

### Account Route (protected)

The `/account` page represents what a user sees when they’re logged into their ChangeBank account. A user that isn’t logged in who tries to access this page should just be taken to the home page.

<Aside type="note">

> In a production application, we would validate the token signature, make sure it wasn't expired, and attempt to refresh it if it were expired.

Author: I have added this PR to verify the token.
https://github.com/FusionAuth/fusionauth-quickstart-javascript-express-web/pull/2

They suggest not to use the package for any automatic refreshing, do you think this is enough?

> In a production application, we would validate the token signature, make sure it wasn't expired, and attempt to refresh it if it were expired.

Reviewer: That's fine, no need to automatically refresh.

faIcon: fa-x
color: fuchsia
coderoot: https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-javascript-express-web/main
cta: EmailListCTA
---
import Aside from '/src/components/Aside.astro';
import LoginAfter from '../../diagrams/quickstarts/login-after.astro';
import LoginBefore from '../../diagrams/quickstarts/login-before.astro';
import RemoteCode from '/src/components/RemoteCode.astro';

In this quickstart you are going to build an application with Express and integrate it with FusionAuth.
You'll be building it for [ChangeBank](https://www.youtube.com/watch?v=CXDxNCzUspM), a global leader in converting dollars into coins.
It'll have areas reserved for users who have logged in as well as public facing sections.

The docker compose file and source code for a complete application are available at
[https://github.com/FusionAuth/fusionauth-quickstart-javascript-express-web](https://github.com/FusionAuth/fusionauth-quickstart-javascript-express-web).

## Prerequisites
* [Node LTS](https://nodejs.org/en/download)
* [Docker](https://www.docker.com): The quickest way to stand up FusionAuth. (There are [other ways](/docs/v1/tech/installation-guide/)).

> * [git](https://git-scm.com/downloads/): Not required but recommended if you want to track your changes.

Reviewer: This is actually not optional, because you tell folks to pull the repo.

```
git clone https://github.com/FusionAuth/fusionauth-quickstart-javascript-express-web.git
cd fusionauth-quickstart-javascript-express-web
```

### Run FusionAuth via Docker

In the root directory of the repo you'll find a Docker compose file (docker-compose.yml) and an environment
variables configuration file (.env). Assuming you have Docker installed on your machine, you can stand up
FusionAuth up on your machine with:

```
docker compose up -d
```

This will start three containers, once each for FusionAuth, Postgres, and Elastic.

Here you are using a bootstrapping feature of FusionAuth, called [Kickstart](/docs/v1/tech/installation-guide/kickstart).
When FusionAuth comes up for the first time, it will look at the `kickstart/kickstart.json` file and
configure FusionAuth to a certain initial state.

<Aside type="note">
If you ever want to reset the FusionAuth system, delete the volumes created by Docker Compose by
executing `docker compose down -v`, then re-run `docker compose up -d`.
</Aside>

FusionAuth will be configured with these settings:
* Your client Id is `e9fdb985-9173-4e01-9d73-ac2d60d1dc8e`.

> * Your client secret is `2HYT86lWSAntc-mvtHLX5XXEpk9ThcqZb4YEh65CLjA-not-for-prod`.

Reviewer: If you aren't going to use the client secret that is common across all quickstarts, please use the `RemoteValue` component: https://github.com/FusionAuth/fusionauth-site/blob/master/astro/src/components/RemoteValue/RemoteValue.astro

Good examples in that file.

* Your example username is `richard@example.com` and the password is `password`.
* Your admin username is `admin@example.com` and the password is `password`.
* The base URL of FusionAuth `http://localhost:9011/`.

You can log into the [FusionAuth admin UI](http://localhost:9011/admin) and look around if you want, but with Docker/Kickstart you don’t need to.

### Setup Application

In this section, you'll set up a basic Express application.

Create a new directory to hold your application, and go into it.

```shell
mkdir changebank
cd changebank
```

Create a file called `package.json` with the below dependencies and scripts. Feel free to update the `author` or other attributes as needed. 

> 

Reviewer: ```suggestion
```

Reviewer: feels a bit jarring otherwise.

</Aside>

### Static files

You need a way to serve your static files, luckily Express makes this easy with `express.static`. These files will include all of the CSS and images that are required for the UI of the ChangeBank application. 

Make sure that you are in the directory `change-bank` when running these commands to create new directories and copy the css and images required from `complete-application`.

```shell
mkdir ../static && mkdir ../static/css && mkdir ../static/img && mkdir ../templates &&
cp ../../complete-application/static/css/changebank.css ../static/css/changebank.css && 
cp ../../complete-application/static/img/money.jpg ../static/img/money.jpg && 
cp ../../complete-application/templates/account.html ../templates/account.html && 
cp ../../complete-application/templates/home.html ../templates/home.html && 
cp ../../complete-application/templates/make-change.html ../templates/make-change.html
```

At the bottom of your `index.ts` file include the below code to serve any assets in the static directory.

<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"

>   tags="static"/>

Reviewer: Do we need a `;` at the end?

### Homepage

In order to show off your homepage you need to create a route with the base URL `/`. In this route check to see if `userDetails` cookie has any user values and correctly respond with either the `home.html` or `account.html` template. Also if there is no user you can generate the PKCE challenge pair and store them into `userSession` cookie (this is an `HTTPOnly` cookie, inaccessible to JavaScript).

<Aside type="note">
In this file you will use the [pkce-challenge](https://www.npmjs.com/package/pkce-challenge) package to generate a Proof Key for Code Exchange (PKCE) challenge pair. Read more about [PKCE](/articles/oauth/modern-guide-to-oauth).
</Aside>

At the bottom of `index.ts` file include the below code, so that the homepage can be displayed when a user is not logged in. If a user is logged in this route will redirect to the `/account` route.

> <RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}

Reviewer: Please make this 2 space indents.

Add the following code to the bottom of `index.ts`.

<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="account"/>

### Make Change

Two functions are needed to accept both `GET` and `POST` requests on the `/make-change` route. The `GET` verifies that the user is logged in, then serves the HTML file required for the page. The `POST` takes the amount form value as a body parameter and uses the it to calculate how many quarters, dimes, nickels and pennies to return. If `total` is not a valid dollar value or cannot be converted an `error` message is returned. The return string `message` is received by asynchronously and updates the `change-message` or `error-message` div accordingly.

Add the following code to the bottom of `index.ts`.

<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="make-change"/>

### Logout Route

<Aside type="tip">
Click the `Logout` button and watch the browser first go to FusionAuth to log out the user, then return to your home page.
</Aside>

> The `/logout` route in your application will redirect to the FusionAuth server. FusionAuth will then end the OAuth session for this user and redirect back to the `logoutURL` specified in your `kickstarter.json` file.

Reviewer: ```suggestion
The `/logout` route in your application will redirect to the FusionAuth server. FusionAuth will then end the FusionAuth SSO session for this user and redirect back to the application's`logoutURL` specified in the `kickstarter.json` file. You'll build that URL handler below.
```

Add the following code to the bottom of `index.ts`.

<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="logout"/>

 The application's logout route is `/oauth2/logout`. Within this route we can safely clear all of the cookies related to this user.

Add the following code to the bottom of `index.ts`.

<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="oauth-logout"/>

### Listen for connection

Listen for connections on port `http://localhost:8080` and serve the Express application.

Add the below code to the bottom of `index.ts`.

<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="app"/>

### Run Application

Return to the root directory of your application `/changebank` by running the below command.

```shell
cd ..
```

Include the necessary environment variables by creating a file called `.env`. Include the following variables in this file.

<RemoteCode url={frontmatter.coderoot + "/complete-application/.env"}
  lang="ts"/>

Now that your application is complete you can run it with the below command.

```shell
npm run dev
```

> 

Reviewer: ```suggestion
You can now open up an incognito window and visit the Rails app at http://localhost:8080/ . Log in with the user account mentioned above, `richard@example.com`.

```

## Next Steps
This quickstart is a great way to get a proof of concept up and running quickly, but to run your application in production, there are some things you're going to want to do.

### FusionAuth Customization
FusionAuth gives you the ability to customize just about everything with the user's experience and your application's integration. This includes
* [Hosted pages](/docs/v1/tech/themes/) such as login, registration, email verification, and many more
* [Email templates](/docs/v1/tech/email-templates/email-templates)

> * [User data and custom claims in access token JWTs](/articles/tokens/jwt-components-explained)

Reviewer: ```suggestion
* [User data and custom claims in access token JWTs](/articles/tokens/jwt-components-explained)

```

### Security
* You may want to customize the [token expiration times and policies](/docs/v1/tech/oauth/#configure-application-oauth-settings) in FusionAuth
* Choose [password rules](/docs/v1/tech/core-concepts/tenants#password) and a [hashing algorithm](/docs/v1/tech/reference/password-hashes) that meet your security needs

> ### Tenant and Application Management

Reviewer: ```suggestion

### Tenant and Application Management
```

---
title: Express
description: Quickstart integration of an Express web application with FusionAuth
navcategory: getting-started
prerequisites: npm, node
section: web
technology: Express
language: JavaScript
icon: /img/icons/javascript.svg
faIcon: fa-x
color: fuchsia
coderoot: https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-javascript-express-web/main
cta: EmailListCTA
---
import Aside from '/src/components/Aside.astro';
import LoginAfter from '../../diagrams/quickstarts/login-after.astro';
import LoginBefore from '../../diagrams/quickstarts/login-before.astro';
import RemoteCode from '/src/components/RemoteCode.astro';

In this quickstart you are going to build an application with Express and integrate it with FusionAuth.
You'll be building it for [ChangeBank](https://www.youtube.com/watch?v=pkH-kD73QUM), a global leader in converting dollars into coins.
It'll have areas reserved for users who have logged in as well as public facing sections.

The docker compose file and source code for a complete application are available at
https://github.com/FusionAuth/fusionauth-quickstart-javascript-express-web

## Prerequisites

> * [Node LTS](https://nodejs.org/en/download)

Reviewer 2: I think we should have a description here and also specify which version we used

* [Docker](https://www.docker.com): The quickest way to stand up FusionAuth. (There are [other ways](/docs/v1/tech/installation-guide/)).
* [git](https://git-scm.com/downloads/): Not required but recommended if you want to track your changes.

## General architecture

While this sample application doesn't have login functionality without FusionAuth, a more typical integration will replace an existing login system with FusionAuth.

In that case, the system might look like this before FusionAuth is introduced.

<LoginBefore alt={"Request flow during login before FusionAuth"}/>

The login flow will look like this after FusionAuth is introduced.

<LoginAfter alt={"Request flow during login after FusionAuth"}/>

In general, you are introducing FusionAuth in order to normalize and consolidate user data. This helps make sure it is consistent and up-to-date as well as offloading your login security and functionality to FusionAuth.

## Getting Started
In this section, you'll get FusionAuth up and running, and configured with the ChangeBank application.

### Clone the Code

> First off, grab the code from the repository and change into that directory.

Reviewer 2: Shouldn't we move this to the end of the article as we're teaching them how to set up everything from scratch?

Author: They will still need the code and FusionAuth running with the kickstart.

```
git clone https://github.com/FusionAuth/fusionauth-quickstart-javascript-express-web.git
cd fusionauth-quickstart-javascript-express-web
```

### Run FusionAuth via Docker

In the root directory of the repo you'll find a Docker compose file (docker-compose.yml) and an environment
variables configuration file (.env). Assuming you have Docker installed on your machine, you can stand up
FusionAuth up on your machine with:

```
docker compose up -d
```

This will start three containers, once each for FusionAuth, Postgres, and Elastic.

Here you are using a bootstrapping feature of FusionAuth, called [Kickstart](/docs/v1/tech/installation-guide/kickstart).
When FusionAuth comes up for the first time, it will look at the `kickstart/kickstart.json` file and
configure FusionAuth to a certain initial state.

<Aside type="note">
If you ever want to reset the FusionAuth system, delete the volumes created by Docker Compose by
executing `docker compose down -v`, then re-run `docker compose up -d`.
</Aside>

> FusionAuth will be configured with these settings

Reviewer 2: ```suggestion
FusionAuth will be configured with these settings:
```

* Your client Id is `e9fdb985-9173-4e01-9d73-ac2d60d1dc8e`.
* Your client secret is `2HYT86lWSAntc-mvtHLX5XXEpk9ThcqZb4YEh65CLjA-not-for-prod`.
* Your example username is `richard@example.com` and the password is `password`.
* Your admin username is `admin@example.com` and the password is `password`.
* The base URL of FusionAuth `http://localhost:9011/`.

You can log into the [FusionAuth admin UI](http://localhost:9011/admin) and look around if you want, but with Docker/Kickstart you don’t need to.

### Setup Application

> In this section, you'll set up a basic Golang application with a single page.

Reviewer 2: ```suggestion
In this section, you'll set up a basic Express application with a single page.
```

  lang="text" />

Create a file called `package.json` with the below dependencies and scripts.

`package.json`
<RemoteCode url={frontmatter.coderoot + "/complete-application/package.json"}
  lang="json" />

Install dependencies using npm.

```shell
npm install
```

Create a new directory to hold your source code, and go into it.

```shell
mkdir src
cd src
```

#### Setup New Application via API

<Aside type="tip">
Although you could also do all of this in the kickstart, this example is meant to show off how to use the SDK on the server for administrative purposes. You could also make any of these changes within the admin UI. You can do this even after the instance is running unlike with kickstart, that needs to happen on a fresh install.
</Aside>

In this setup you will:

1. Update the Tenant's issuer to `http://localhost:9011`
1. Generate an RSA Key for the App
1. Create a new Application named `JSExampleApp` using the new RSA key.
1. Find the User with email `richard@example.com`
1. Update `richard@example.com`'s fullName property to `Richard Bray`
1. Register the User `Richard Bray` with the new application

> Make sure you are still in the directory `complete-application/src`

Reviewer 2: ```suggestion
Make sure you are still in the directory `changebank/src`
```

The instructions said we should've changed into `changebank`

Create a file called `setup.ts` with the below code.

`setup.ts`
<RemoteCode url={frontmatter.coderoot + "/complete-application/src/setup.ts"}
  lang="ts" />

Execute this setup by calling an npm script.

```shell
npm run setup
```

## Create the Express Application

> In order for our application to look good we have some files that are already made, copy `complete-application/static` and `complete-application/templates` to the `changebank` directory. These static files are used to create the "beautiful" Changebank application UI.

Reviewer 2: ```suggestion
In order for our application to look good we have some files that are already made, copy `complete-application/static` and `complete-application/templates` to the `changebank` directory. These static files are used to create the "beautiful" ChangeBank application UI.
```

<Aside type="note">
You can choose to examine the full example of the running code now, or skip below and walk through each step of the process.
</Aside>

### Full Example

> Make sure once again that you are in the directory `complete-application/src`.

Reviewer 2: But if they cloned the repository, they already have this file with these contents, right?

Author: removed the creation portion of this, it was duplicate from below. The Full Example is the `complete-application` already set to run.

Create a file called `index.ts` with the below code.

`index.ts`
<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts" />

#### Run the App!
You should now be able to start your Express application with 

```shell
npm run dev
```

### Walkthrough Example

> Make sure once again that you are in the directory `complete-application/src`.

Reviewer 2: Shouldn't we go back to `changebank/src` again?

Author: yes

Create a file called `index.ts` and start by setting up your imports and constants. These constants have already been defined by running the kickstart and `setup.js`.

`index.ts`
<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="top"/>

<Aside type="note">
If you want to see where the FusionAuth values came from, they can be found in the 
FusionAuth app (http://localhost:9011/admin). The tenant ID is found on the Tenants

> page. To see the client ID and client secret, go to the Applications page and 

Reviewer 2: ```suggestion
page. To see the Client Id and Client Secret, go to the Applications page and 
```

> click the `View` icon under the actions for the Changebank application. You'll 

Reviewer 2: ```suggestion
click the `View` icon under the actions for the ChangeBank application. You'll 
```

> find the client id and client secret values in the `OAuth configuration` section.

Reviewer 2: ```suggestion
find the Client Id and Client Secret values in the `OAuth configuration` section.
```

</Aside>

#### Static files

We need a way to serve our static files, luckily Express makes this easy with `express.static`. These files will include all of the CSS and images that are required for the UI of the ChangeBank application.

`index.ts`
<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="static"/>

#### Homepage

> In order to show off our homepage we need to create a route with the base url `/`. In this route check to see if `userDetails` cookie has any user values and correctly respond with either the `home.html` or `account.html` template. Also if there is no user you can generate the PKCE challenge pair and store them into `userSession` cookie (take note this is an http-only cookie).

Reviewer 2: ```suggestion
In order to show off our homepage, we need to create a route with the base URL `/`. In this route, check to see if the `userDetails` cookie has any user values and correctly respond with either the `home.html` or `account.html` template. Also, if there is no user you can generate the PKCE challenge pair and store them into `userSession` cookie (take note this is an HTTP-Only cookie).
```

I don't know whether these suggestions are grammatically correct, but I felt we needed some commas for a better flow.

<Aside type="note">

> In this file we will use the package [pkce-challenge](https://www.npmjs.com/package/pkce-challenge) to generate a Proof Key for Code Exchange (PKCE) challenge pair. Read more about [PKCE](https://www.oauth.com/oauth2-servers/pkce/authorization-request/).

Reviewer 2: ```suggestion
In this file we will use the package [pkce-challenge](https://www.npmjs.com/package/pkce-challenge) to generate a Proof Key for Code Exchange (PKCE) challenge pair. Read more about [PKCE](/articles/oauth/modern-guide-to-oauth).
```

</Aside>

`index.ts`
<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="homepage"/>

#### Login Route

> In the Changebank application there is a `Login` button, when clicked will open the `/login` route (you can find this in `complete-application/templates/home.html`). Have this redirect them to FusionAuth so that FusionAuth can present them with a login page.

Reviewer 2: ```suggestion
In the ChangeBank application there is a `Login` button which will open the `/login` route when clicked (you can find this in `complete-application/templates/home.html`). Have this redirect them to FusionAuth so that FusionAuth can present them with a login page.
```

I think that the second sentence is kind of weird

`index.ts`
<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="login"/>

> You can always find this url if you view the details of your Application within FusionAuth, under "OAuth2 & OpenID Connect Integration details" you will find the property `OAuth IdP login URL`. The endpoint that you will hit on FusionAuth is `/oauth2/authorize` below you can find a list of the needed parameters for FusionAuth's [OAuth Endpoints](https://fusionauth.io/docs/v1/tech/oauth/endpoints)

Reviewer 2: I think we should rephrase this paragraph, it's kind of confusing

Author: How about

The OAuth IdP login URL can be found in the "OAuth2 & OpenID Connect Integration details" section of your application's settings in FusionAuth. The endpoint that you will hit on FusionAuth is /oauth2/authorize. You can find a list of the required parameters for FusionAuth's OAuth endpoints here: [OAuth Endpoints](/docs/v1/tech/oauth/endpoints)

Reviewer 2: Awesome!

> You can always find this url if you view the details of your Application within FusionAuth, under "OAuth2 & OpenID Connect Integration details" you will find the property `OAuth IdP login URL`. The endpoint that you will hit on FusionAuth is `/oauth2/authorize` below you can find a list of the needed parameters for FusionAuth's [OAuth Endpoints](https://fusionauth.io/docs/v1/tech/oauth/endpoints)

Reviewer 2: ```suggestion
You can always find this URL if you view the details of your Application within FusionAuth, under "OAuth2 & OpenID Connect Integration details" you will find the property `OAuth IdP login URL`. The endpoint that you will hit on FusionAuth is `/oauth2/authorize` below you can find a list of the needed parameters for FusionAuth's [OAuth Endpoints](/docs/v1/tech/oauth/endpoints)
```

Author: addressed in new paragraph suggestion

> - client_id: Your `applicationId`
> - response_type: `code` since we are using the code flow
> - redirect_uri: Where your FusionAuth will redirect back to during the Oauth flow, in our example this is our route `oauth-redirect`
> - state: Created in our homepage route using some crazy math and stored in our cookie as `stateValue`
> - code_challenge: Created in our homepage route using `pkce-challenge` stored in our cookie as `challenge`, this is one half of the PKCE 

Reviewer 2: ```suggestion
- `client_id`: Your `applicationId`
- `response_type`: `code` since we are using the Authorization Code flow
- `redirect_uri`: Where your FusionAuth will redirect back to during the OAuth flow (in our example, this is our route `oauth-redirect`)
- `state`: Random string created in our homepage route and stored in our cookie as `stateValue`
- `code_challenge`: Created in our homepage route using `pkce-challenge` and stored in our cookie as `challenge`, this is one half of the PKCE 
```

#### OAuth Redirect

> When FusionAuth redirects back to our Express application it will return to our `/oauth-redirect` route. In this response it will have two query parameters that we need to further examine `code` and `state`. 

Reviewer 2: ```suggestion
When FusionAuth redirects back to our Express application, it will return to our `/oauth-redirect` route. In this response it will have two query parameters that we need to further examine: `code` and `state`. 
```

Not sure about commas again

1. Make sure to check that the `stateValue` that we stored in our cookie matches the `state` query parameters
1. Request to the FusionAuth Token endpoint to exchange the authorization `code` returned from the Authorize endpoint and a `verifier` from our `userSession` cookie for an access token.

> 1. Retrieves the user by passing the access token to FusionAuth. (This method does not use an API key, instead it uses a JSON Web Token (JWT) for authentication)
> 1. Store access_token details in `userToken` cookie.

Reviewer 2: ```suggestion
1. Retrieve the user by passing the access token to FusionAuth. (This method does not use an API key, instead it uses a JSON Web Token (JWT) for authentication).
1. Store access token details in `userToken` cookie.
```

1. Store retrieved user in `userDetails` cookie (note for our example this is not a secure cookie and JS can access).
1. Redirect to `/account` route.

`index.ts`

> <RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}

Reviewer 2: I think we should use [`timingSafeEqual()`](https://nodejs.org/api/crypto.html#cryptotimingsafeequala-b) instead of `if (stateFromFusionAuth !== userSessionCookie?.stateValue) {`

Author: This is a string comparison, how does this work for that?

Reviewer 2: It's just a time-constant string comparison, so we could do something like this to avoid leaking information about the state string to an attacker. But maybe I'm being overzealous.

```ts
if (!timingSafeEqual(stateFromFusionAuth, userSessionCookie?.stateValue)) {
```

Author: This seems overkill to me as it will be on the server.

  lang="ts"
  tags="oauth-redirect"/>

#### Account Route (protected)

> The /account page represents what a user sees when they’re logged into their Changebank account. A user that isn’t logged in, who tries to access this page, should just be taken to the home page.

Reviewer 2: ```suggestion
The `/account` page represents what a user sees when they’re logged into their ChangeBank account. A user that isn’t logged in who tries to access this page should just be taken to the home page.
```

I'm not sure about removing those commas but it felt better.

<Aside type="note">
In a production application, we would validate the token signature, make sure it wasn't expired, and attempt to refresh it if it were expired.
</Aside>

`index.ts`
<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="account"/>

#### Logout Route

<Aside type="tip">
Click the `Logout` button and watch the browser first go to FusionAuth to log out the user, then return to your home page.
</Aside>

The last step is to implement logout. When you log a user out of an application, you'll take them to 
FusionAuth's `/oauth2/logout` endpoint. After logging the user out, FusionAuth will redirect the user
to your app's `/logout` endpoint, which you'll create now. This endpoint deletes any cookies that your
application created, clearing the user's session.

`index.ts`
<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
  lang="ts"
  tags="logout"/>

> In the below code we check to see if the current user session has any information about the user, if not we redirect to the FusionAuth server to initiate the logout on the server. When this returns the user session cookie will now be empty and we can therefore forward back to our homepage, if we didn't add this check we would be stuck in an infinite loop between our Express server and FusionAuth (luckily unlike other providers you won't get billed per all :D)

Reviewer 2: ```suggestion
In the below code, we check if the current user session has any information about the user and redirect to the FusionAuth server to initiate the logout on the server. When this returns, the user session cookie will now be empty and we can therefore forward back to our homepage. If we didn't add this check, we would be stuck in an infinite loop between our Express server and FusionAuth.
```

I removed the last part but am okay to leave it (with no emoji).

```ts
if (userSession) {
    res.redirect(302, `${fusionAuthURL}/oauth2/logout?client_id=${clientId}`);
} else {
    res.redirect(302, '/')
}
```

## Bonus
<Aside type="tip">

> As a bonus, try implementing the `/make-change` route using the `complete-application/templates/make-change.html` template, implementing the same validation as `/account` route.

Reviewer 2: I thought we'd have to implement this in our quickstarts

Reviewer 2: Great! But please don't forget to add those steps here.

Author: added if you would like to review

</Aside>

## Next Steps
This quickstart is a great way to get a proof of concept up and running quickly, but to run your application in production, there are some things you're going to want to do.

> ### FusionAuth Customization
> FusionAuth gives you the ability to customize just about everything with the user's experience and your application's integration. This includes
> * [Hosted pages](https://fusionauth.io/docs/v1/tech/themes/) such as login, registration, email verification, and many more
> * [Email templates](https://fusionauth.io/docs/v1/tech/email-templates/email-templates)
> * [User data and custom claims in access token JWTs](https://fusionauth.io/articles/tokens/jwt-components-explained)
> 
> ### Security
> * You may want to customize the [token expiration times and policies](https://fusionauth.io/docs/v1/tech/oauth/#configure-application-oauth-settings) in FusionAuth
> * Choose [password rules](https://fusionauth.io/docs/v1/tech/core-concepts/tenants#password) and a [hashing algorithm](https://fusionauth.io/docs/v1/tech/reference/password-hashes) that meet your security needs
> 
> ### Tenant and Application Management
> * Model your application topology using [Applications](https://fusionauth.io/docs/v1/tech/core-concepts/applications), [Roles](https://fusionauth.io/docs/v1/tech/core-concepts/roles), [Groups](https://fusionauth.io/docs/v1/tech/core-concepts/groups), [Entities](https://fusionauth.io/docs/v1/tech/core-concepts/groups), and more
> * Set up [MFA](https://fusionauth.io/docs/v1/tech/guides/multi-factor-authentication), [Social login](https://fusionauth.io/docs/v1/tech/identity-providers/), and/or [SAML](https://fusionauth.io/docs/v1/tech/identity-providers/samlv2/) integrations
> * Integrate with external systems using [Webhooks](https://fusionauth.io/docs/v1/tech/events-webhooks/), [SCIM](https://fusionauth.io/docs/v1/tech/core-concepts/scim), and [Lambdas](https://fusionauth.io/docs/v1/tech/lambdas/)

Reviewer 2: ```suggestion
### FusionAuth Customization
FusionAuth gives you the ability to customize just about everything with the user's experience and your application's integration. This includes
* [Hosted pages](/docs/v1/tech/themes/) such as login, registration, email verification, and many more
* [Email templates](/docs/v1/tech/email-templates/email-templates)
* [User data and custom claims in access token JWTs](/articles/tokens/jwt-components-explained)

### Security
* You may want to customize the [token expiration times and policies](/docs/v1/tech/oauth/#configure-application-oauth-settings) in FusionAuth
* Choose [password rules](/docs/v1/tech/core-concepts/tenants#password) and a [hashing algorithm](/docs/v1/tech/reference/password-hashes) that meet your security needs

### Tenant and Application Management
* Model your application topology using [Applications](/docs/v1/tech/core-concepts/applications), [Roles](/docs/v1/tech/core-concepts/roles), [Groups](/docs/v1/tech/core-concepts/groups), [Entities](/docs/v1/tech/core-concepts/groups), and more
* Set up [MFA](/docs/v1/tech/guides/multi-factor-authentication), [Social login](/docs/v1/tech/identity-providers/), and/or [SAML](/docs/v1/tech/identity-providers/samlv2/) integrations
* Integrate with external systems using [Webhooks](/docs/v1/tech/events-webhooks/), [SCIM](/docs/v1/tech/core-concepts/scim), and [Lambdas](/docs/v1/tech/lambdas/)
```

Just removing absolute links
